tlog_param.py

```diff
@@ -1,39 +1,97 @@
+
+import pymavlink.dialects.v20.common as mav_common
 from pymavlink import mavutil
+
 import util
 
 # Use MAVLink2
 os.environ['MAVLINK20'] = '1'
 
 
+def firmware_version_type_str(firmware_version_type: int) -> str:
+    if firmware_version_type == mav_common.FIRMWARE_VERSION_TYPE_DEV:
+        return 'dev'
+    elif firmware_version_type == mav_common.FIRMWARE_VERSION_TYPE_ALPHA:
+        return 'alpha'
+    elif firmware_version_type == mav_common.FIRMWARE_VERSION_TYPE_BETA:
+        return 'beta'
+    elif firmware_version_type == mav_common.FIRMWARE_VERSION_TYPE_RC:
+        return 'rc'
+    else:
+        return ''
+
+
+class Param(NamedTuple):
+    value: str
+    type: str   # Actual values are MAV_PARAM_TYPE_*, but we'll store them as strings
+
+
 class TelemetryLogParam:
-    def __init__(self, tlog_filename: str):
-        self.tlog_filename = tlog_filename
+    def __init__(self):
+        self.params: dict[str, Param] = {}
+        self.autopilot_version: str = ''
+        self.git_hash: str = ''
 
-    def read_and_report(self):
-        print(f'Results for {self.tlog_filename}')
-        mlog = mavutil.mavlink_connection(self.tlog_filename, robust_parsing=False, dialect='ardupilotmega')
+    def read(self, infile: str):
+        mlog = mavutil.mavlink_connection(infile, robust_parsing=False, dialect='ardupilotmega')
 
-        params = {}
-        while (msg := mlog.recv_match(blocking=False, type=['PARAM_VALUE'])) is not None:
+        while (msg := mlog.recv_match(blocking=False, type=['PARAM_VALUE', 'AUTOPILOT_VERSION'])) is not None:
             data = msg.to_dict()
-            params[data['param_id']] = data['param_value']
+            if msg.get_type() == 'PARAM_VALUE':
+                param_id = data['param_id']
+                param_value = data['param_value']
 
-        # Print results
-        for pi in sorted(params.items()):
-            print(f'{pi[0]} = {pi[1]}')
+                if param_id in self.params and self.params[param_id].value != param_value:
+                    print(f'{param_id} was {self.params[param_id].value}, changed to {param_value}')
 
+                self.params[param_id] = Param(param_value, data['param_type'])
+            else:
+                flight_sw_version = data['flight_sw_version']
+                major = (flight_sw_version >> (8 * 3)) & 0xFF
+                minor = (flight_sw_version >> (8 * 2)) & 0xFF
+                path = (flight_sw_version >> (8 * 1)) & 0xFF
+                version_type = (flight_sw_version >> (8 * 0)) & 0xFF
+
+                self.autopilot_version = f'{major}.{minor}.{path} {firmware_version_type_str(version_type)}'
+                self.git_hash = bytes(data['flight_custom_version']).decode('utf-8')
+
+    def write(self, outfile: str):
+        """
+        Write a QGC-compatible params file
+
+        File format: https://dev.qgroundcontrol.com/master/en/file_formats/parameters.html
+        """
+        if len(self.params) == 0:
```

Comment:
```
if not self.params:
```

is a bit more Pythonic, assuming None, '', and {} all mean there's nothing to print.

Reply:
Done